pkg/datapath/linux/ipsec/ipsec_linux.go

```diff
@@ -149,6 +149,96 @@ func ipSecJoinState(state *netlink.XfrmState, keys *ipSecKey) {
 	state.Reqid = keys.ReqID
 }
 
+// xfrmStateReplace attempts to add a new XFRM state only if one doesn't
+// already exist. If it doesn't but some other XFRM state conflicts, then
+// we attempt to remove the conflicting state before trying to add again.
+func xfrmStateReplace(new *netlink.XfrmState) error {
+	states, err := netlink.XfrmStateList(netlink.FAMILY_ALL)
+	if err != nil {
+		return fmt.Errorf("Cannot get XFRM state: %s", err)
+	}
+
+	// Check if the XFRM state already exists
+	for _, s := range states {
+		if xfrmIPEqual(s.Src, new.Src) && xfrmIPEqual(s.Dst, new.Dst) &&
+			xfrmMarkEqual(s.OutputMark, new.OutputMark) &&
+			xfrmMarkEqual(s.Mark, new.Mark) && s.Spi == new.Spi {
+			return nil
+		}
+	}
+
+	// It doesn't exist so let's attempt to add it.
+	firstAttemptErr := netlink.XfrmStateAdd(new)
+	if !os.IsExist(firstAttemptErr) {
+		return firstAttemptErr
+	}
+	log.WithFields(logrus.Fields{
+		logfields.SPI:           new.Spi,
+		logfields.SourceIP:      new.Src,
+		logfields.DestinationIP: new.Dst,
+	}).Warn("Failed to add XFRM state due to conflicting state")
+
+	// An existing state conflicts with this one. We need to remove the
+	// existing one first.
+	deletedSomething, err := xfrmDeleteConflictingState(states, new)
+	if err != nil {
+		return err
+	}
+
+	// If no conflicting state was found and deleted, there's no point in
+	// attempting to add again.
+	if !deletedSomething {
+		return firstAttemptErr
+	}
+	return netlink.XfrmStateAdd(new)
+}
+
+// Attempt to remove any XFRM state that conflicts with the state we just tried
+// to add. To find those conflicting states, we need to use the same logic that
+// the kernel used to reject our check with EEXIST. That logic is upstream in
+// __xfrm_state_lookup.
+func xfrmDeleteConflictingState(states []netlink.XfrmState, new *netlink.XfrmState) (bool, error) {
+	deletedSomething := false
+	for _, s := range states {
+		if new.Spi == s.Spi && new.Mark == s.Mark &&
```

Comment:
`xfrmState.Mark` is of type `*XfrmMark`, so `new.Mark == s.Mark` is true only if `new.Mark == s.Mark == nil`, or they point to the same object, but the latter one can't be true, am I right?